gtsfm/frontend/cacher/matcher_cacher.py

```diff
@@ -0,0 +1,191 @@
+                descriptors_i2=descriptors_i2,
+                im_shape_i1=im_shape_i1,
+                im_shape_i2=im_shape_i2,
+            )
+        )
+
+        return io_utils.read_from_compressed_file(cache_path)
+
+    def __save_result_to_cache(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+        match_indices: np.ndarray,
+    ) -> None:
+        """Save the results (match indice) to the cache."""
+        cache_path = self.__get_cache_path(
+            cache_key=self.__generate_cache_key(
+                keypoints_i1=keypoints_i1,
+                keypoints_i2=keypoints_i2,
+                descriptors_i1=descriptors_i1,
+                descriptors_i2=descriptors_i2,
+                im_shape_i1=im_shape_i1,
+                im_shape_i2=im_shape_i2,
+            )
+        )
+        io_utils.write_to_compressed_file(match_indices, cache_path)
+
+    def match(
+        self,
+        keypoints_i1: Keypoints,
+        keypoints_i2: Keypoints,
+        descriptors_i1: np.ndarray,
+        descriptors_i2: np.ndarray,
+        im_shape_i1: Tuple[int, int],
+        im_shape_i2: Tuple[int, int],
+    ) -> np.ndarray:
+        """Match descriptor vectors.
+
+        If the results are in the cache, they are fetched and returned. Otherwise, the `detect_and_describe` of the
+        underlying object's API is called and the results are cached.
+
+        # Some matcher implementations (such as SuperGlue) utilize keypoint coordinates as
+        # positional encoding, so our matcher API provides them for optional use.
+
+        Output format:
+        1. Each row represents a match.
+        2. First column represents keypoint index from image #i1.
+        3. Second column represents keypoint index from image #i2.
+        4. Matches are sorted in descending order of the confidence (score), if possible.
+
+        Args:
+            keypoints_i1: keypoints for image #i1, of length N1.
+            keypoints_i2: keypoints for image #i2, of length N2.
+            descriptors_i1: descriptors corr. to keypoints_i1.
+            descriptors_i2: descriptors corr. to keypoints_i2.
+            im_shape_i1: shape of image #i1, as (height,width).
+            im_shape_i2: shape of image #i2, as (height,width).
+
+
+        Returns:
+            Match indices (sorted by confidence), as matrix of shape (N, 2), where N < min(N1, N2).
+        """
+        cached_data = self.__load_result_from_cache(
+            keypoints_i1=keypoints_i1,
+            keypoints_i2=keypoints_i2,
+            descriptors_i1=descriptors_i1,
+            descriptors_i2=descriptors_i2,
+            im_shape_i1=im_shape_i1,
+            im_shape_i2=im_shape_i2,
+        )
+
+        if cached_data is not None:
+            return cached_data
```

Comment:
maybe a print statement here as well to inform user that cached results are being used to return `match_indices`. Although it may too many print statements for all the image pairs involved in two view matching.

Reply:
I like this idea too